### ISSUE TITLE
[Feature Request] Adding bookmarklet support

### ISSUE THREAD

**Issue by wolnavi** [duckduckgo/iOS#694]
Hi. is it possible to add bookmarklet support? Currently, when adding a bookmarklet, it will always prefix `http://`, I'm not a real coder but perhaps there should be something in this file

https://github.com/duckduckgo/iOS/blob/develop/DuckDuckGo/EditBookmarkAlert.swift

at line 83

where current code is 

````swift
if !urlString.hasPrefix("http://") && !urlString.hasPrefix("https://") {
                urlString = "http://\(urlString)"
            }
````

So we can replace with something like

````swift
if !urlString.hasPrefix("javascript:") {
                urlString = "javascript:\(urlString)"
            } else if !urlString.hasPrefix("http://") && !urlString.hasPrefix("https://") {
                urlString = "http://\(urlString)"
            }
````

Or however it's done... Thanks

**Comment by brindy** [comment]
Hi @wolnavi - I ran this past the team today and we couldn't think of any strong objections.  If you'd like to make a PR for this we can review or we'll try and get it in when we have chance. Thanks!

**Comment by ivantpapr2018** [comment]
> Hi @wolnavi - I ran this past the team today and we couldn't think of any strong objections. If you'd like to make a PR for this we can review or we'll try and get it in when we have chance. Thanks!

Like I said, I’m not a real coder, just guessing around, I don’t even have a Mac, only iPhone 😂😂

**Comment by wolnavi** [comment]
> Hi @wolnavi - I ran this past the team today and we couldn't think of any strong objections. If you'd like to make a PR for this we can review or we'll try and get it in when we have chance. Thanks!

Like I said, I’m not a real coder, just guessing around, I don’t even have a Mac, only a iPhone 😂😂

**Comment by brindy** [comment]
Thanks for the PR @jbttn - tentatively, it looks good, but I need to open a product review request internally to get it approved.  Please bear with me.  Tracking this internally at https://app.asana.com/0/392891325557410/1195769562482248

**Comment by brindy** [comment]
@wolnavi and @jbttn - can you describe some use cases where this would be useful, please?  

We don't support `prompt` or `alert` so it feels as though this is fairly limited in its usefulness.

**Comment by wolnavi** [comment]
> @wolnavi and @jbttn - can you describe some use cases where this would be useful, please?
> 
> We don't support `prompt` or `alert` so it feels as though this is fairly limited in its usefulness.

I don't understand what you mean by `prompt` or `alert` because the current way to edit bookmarks will display a `prompt` or `alert`

As to usefulness, I'm not really sure how to answer this but haven't you used a bookmarklet before?
Quick google:

> A bookmarklet is a bookmark stored in a web browser that contains JavaScript commands that add new features to the browser. Bookmarklets are unobtrusive JavaScripts stored as the URL of a bookmark in a web browser or as a hyperlink on a web page. Bookmarklets are usually JavaScript programs.

As to why would I want this feature, for example, if we visit a webpage with ads, I could add a bookmarklet to remove the ads via javascript and bookmarklets will allow me to execute that

**Comment by brindy** [comment]
> I don't understand what you mean by prompt or alert because the current way to edit bookmarks will display a prompt or alert

I mean the JavaScript `prompt` and `alert` functions.  `prompt` at least seems to be commonly used in bookmarklets.

> As to why would I want this feature, for example, if we visit a webpage with ads, I could add a bookmarklet to remove the ads via javascript and bookmarklets will allow me to execute that

Alright, thanks.  I just needed a concrete example.

**Comment by jbttn** [comment]
@wolnavi Regarding the `prompt` and `alert` confusion...

1) I used `javascript:(function() { alert(1) })()` in the unit test code for this pull request.  This is only meant for the unit tests, it was the simplest thing I could think of to use.
2) I referenced [this page](https://caiorss.github.io/bookmarklet-maker/) for some example bookmarklets.  I didn't realize so many of them used `prompt`.

**Comment by brindy** [comment]
Just FYI, I am trying to progress this internally.  My feeling right now is that we won't proceed with it without the `alert` and `prompt` functionality and they need deep thought around UX (think about pages spamming the alert dialog and how to handle that in an elegant way), but I am making the case for it, so please bear with me.

**Comment by wolnavi** [comment]
> Just FYI, I am trying to progress this internally. My feeling right now is that we won't proceed with it without the `alert` and `prompt` functionality and they need deep thought around UX (think about pages spamming the alert dialog and how to handle that in an elegant way), but I am making the case for it, so please bear with me.

Actually I don't understand, bookmarklet is supported on almost every browser, shouldn't it be a basic functionality? 

Regarding the UX for pages spamming the alert dialog, I don't think a webpage will affect and interfere with the bookmarklet, as I believe the bookmarklet is user defined, so unless s/he wants to spam him/herself with an infinite loop of alerts, we usually won't face this scenario in real world usage...

**Comment by brindy** [comment]
This has been done and is in the release that is currently rolling out. (7.56.0)

Thanks @wolnavi for opening this issue and @jbttn for your contribution.

**Comment by wolnavi** [comment]
This doesn’t seem to work as of Version 7.57.0 ? Adding a bookmarklet to execute an alert does not prompt any alert

**Comment by jbttn** [comment]
@wolnavi as @brindy [mentioned here](https://github.com/duckduckgo/iOS/issues/694#issuecomment-698861988) the app does not support prompt or alert.  I can confirm bookmarklets do work, it's just that any calls to `alert` or `prompt` in the bookmarklet will do nothing.

**Comment by wolnavi** [comment]
> @wolnavi as @brindy [mentioned here](https://github.com/duckduckgo/iOS/issues/694#issuecomment-698861988) the app does not support prompt or alert. I can confirm bookmarklets do work, it's just that any calls to `alert` or `prompt` in the bookmarklet will do nothing.

Trying:

````javascript
javascript:currentUrl=document.location.href.split('?')%5B0%5D.slice(0);document.location.assign(currentUrl+'images');
````

On `google.com` AS AN EXAMPLE that I can currently think of now, does not work, although it works on safari

**Comment by brindy** [comment]
Not sure what that is supposed to do (it just seems to add the word `images` to the current url) but it works for me in both Safari and DDG.

**Comment by wolnavi** [comment]
> Not sure what that is supposed to do (it just seems to add the word `images` to the current url) but it works for me in both Safari and DDG.

Yes, I just want to demonstrate if bookmarklet is working, since alert function is disabled in DDG browser 

As shown in this recording https://youtu.be/ra_W_pilVp0

I first inspect bookmarklet to show that they are the same, then I try to tap on it, but it does nothing (it stays on google search and does not redirect to google images)

Later on, I go to Safari, then inspect bookmarklet to show that they are the same, then I try to tap on it and it works (you can see it redirect to google images from google search)

I reinstalled DDG but it still doesn’t work.. I am on iOS 12.4.1 if that matters

**Comment by bwaresiak** [comment]
@wolnavi I was able to reproduce your problem - it is related to using Open In External Apps option from settings. 

Fix in https://github.com/duckduckgo/iOS/pull/744 In the meantime you may disable that option and Bookmarklet should work.

**Comment by wolnavi** [comment]
> @wolnavi I was able to reproduce your problem - it is related to using Open In External Apps option from settings.
> 
> Fix in #744 In the meantime you may disable that option and Bookmarklet should work.

Ooohh yes it works now, thank you

**Comment by wolnavi** [comment]
Fixed as of v7.57.1, thanks everyone!

**Comment by nonnorm** [comment]
Just out of curiosity, why do you not support `prompt`/`alert`?  Is it because of the possibility of pages being able to spam them?  I would like it to be implemented, but I'm sure there's a good reason it hasn't yet.
